Enables URL parsing of p5.js version and add-on libraries

Hugo-RM opened the pull request:
Fixes #3647

## Changes

- Adds URL parser to allow p5.js version to be configured from URL.
- Includes URL parser test file to test edge cases and fallback response.

### Potential future extensions addressed

- Supports latest patch and minor versions as requested in issue #3647
- URL parser allows for toggling on and off add-on libraries with robust variants allowed.

## Demo Video

[Demo Video](https://youtu.be/Kg6xKqNQFTQ) (Wasn't able to upload it directly sorry!)

Video demonstrates:
- Default behavior (remains the same)
- Different version inputs
- Add-on toggling functionality
- Fallback behavior for invalid inputs
- Edge case handling for the sound add-on URL with p5 version 2.x vs 1.x/0.x

## Note about TypeScript migration

We began working on this before contributions to the migration effort were open to the public (October 31, 2025). As such, our implementation is in JavaScript. Given that it is now past October 31, we can:

- Edit our changes to utilize TypeScript
- Open a separate issue for migrating the URL parser to TypeScript
- Leave the changes as is
- Another approach (your choice)

Let us know your preference.

## I have verified that this pull request:

* [x] has no linting errors (`npm run lint`)
* [x] has no test errors (`npm run test`)
* [x] is from a uniquely-named feature branch and is up to date with the  `develop` branch.
* [x] is descriptively named and links to an issue number, i.e. `Fixes #123`
* [x] meets the standards outlined in the [accessibility guidelines](https://github.com/processing/p5.js-web-editor/blob/develop/contributor_docs/accessibility.md)

Hugo-RM committed:
Added unfinished URL parser and helper functions

Helper functions will ensure URL parameters are
normalized before being used.

Hugo-RM committed:
Refactor defaultHTML to accept options object

Convert defaultHTML from a constant string to a function that accepts
an options object for version and add-on library configuration. This
enables URL parameter parsing for library selection.

Maintains backward compatibility - calling defaultHTML() with no
arguments produces the same output as before.

Enochteo committed:
Added test file: client/utils/parseURLParams.test.js
I have added a test file for the URL parameter parsing utility located at client/utils/parseURLParams.js. This test file includes various test cases to ensure the correct functionality of the parseURLParams function, covering edge cases and typical usage scenarios. For the add-on flags I allowed 'on', 'true', '1', 'ON', 'True', and 'TRUE' to be interpreted as true, while 'off', 'false', '0', 'OFF', 'False', and 'FALSE' are interpreted as false. This should enhance the robustness of the URL parameter parsing in our application.

GoodKimchi committed:
added functions, defaults,

Hugo-RM committed:
Finish validator functions for URL parser

Add validation functions to ensure URL parameters for p5.js versioning
is correctly parsed and handled. Falls back to default functionality
for invalid/missing parameters.

Use npx prettier --write to format all modified files.

Change test fallbacks to be consistent with default behavior.

Add URL parser to client/modules/IDE/reducers/file.js

Co-authored-by: Oscar Bedolla <GoodKimchi@users.noreply.github.com>
Co-authored-by: Enoch Owoade <enochowoade@gmail.com>

Hugo-RM committed:
Merge branch 'develop' into feat/url-parsing-p5.js-version

raclim committed:
Merge branch 'develop' into feat/url-parsing-p5.js-version

raclim commented on server/domain-objects/createDefaultFiles.js:
```diff
@@ -8,11 +8,35 @@ function draw() {
 }`;
 
-export const defaultHTML = `<!DOCTYPE html>
+export function defaultHTML({
```
This is a great change—it adds much more flexibility in creating the default files and the changes here are easy to follow!

raclim commented:
Thanks so much for your work and patience @Hugo-RM and @Enochteo! I'm sorry about the delay in getting back on this! 

I really appreciate the thorough documentation and clarity of your process—I feel that it conveys a clear understanding of the issue, potential edge cases, and where it fits in relation to other ongoing work. I think the proposed solution overall works really well by adding in the URL parsing functionality in a simple and consistent way, and is most likely good to go! 

Regarding refactoring it to TypeScript, I think going with the second option of "opening a separate issue for migrating the URL parser to TypeScript" would be best! Please feel free to go ahead with opening a new issue for it if that would be something the both of you would be interested in! 

Tagging @ksen0 in case you have any other thoughts here—also linking [a URL to a Release environment for this PR](https://app-tedcacb.release.editor.p5js.org/) to quickly test it out!

ksen0 commented:
Thanks so much for this thorough and well=documented work @Hugo-RM and @Enochteo! I also really appreciated the video.

I noticed that the addon flags (I tried to follow how the video presented it) did not work for me:

* https://app-tedcacb.release.editor.p5js.org/?version=2.1.1&data=ON
* https://app-tedcacb.release.editor.p5js.org/?version=2.1.1&data=on
* https://app-tedcacb.release.editor.p5js.org/?version=2.1.1&data=1

all still have the default off on data for me. If I've misunderstood the usage, how should it look like instead?
In any case, really great feature, excited to see it live!

Hugo-RM commented:
Thank you for the kind words!

@ksen0 I apologize for the issue, you were using it correctly. There is an issue I missed.

### Issue

After looking through the recently merged commits, I realized that the issue with the data & events add-on stems from the recent change from version 0.1.2 to 0.2.0 in their URLs. This means preload and shapes have also stopped working.

#### `server/domain-objects/createDefaultFiles.js`
```.js
    preload                                               // Should be 0.2.0 now
      ? `<script src="https://cdn.jsdelivr.net/npm/p5.js-compatibility@0.1.2/src/preload.js"></script>`
      : '',
    shapes                                                // Should be 0.2.0 now
      ? `<script src="https://cdn.jsdelivr.net/npm/p5.js-compatibility@0.1.2/src/shapes.js"></script>`
      : '',
    data                                                  // Should be 0.2.0 now
      ? `<script src="https://cdn.jsdelivr.net/npm/p5.js-compatibility@0.1.2/src/data.js"></script>`
```

### Testing

I tested this by simply switching the version in `server/domain-objects/createDefaultFiles.js`. URLs with the outdated versions would not work but those with updated ones would. 

### Proposed Fix

I see two possible fixes but would like to ask which would be preferred.

- Move the URLs to `common/` and export it from there.
  - I imagine this as similar to how `currentP5Version` is implemented and used currently.
  - Changes apply to all usages of the variables.
- Update the URL strings in `createDefaultFiles.js` to use 0.2.0 (temporary fix, not ideal imo).

I believe importing the URLs from `useP5Version.jsx` into `createDefaultFiles.js` would not be possible because a server file is importing from a client file. If this is not an issue, I can try doing this instead if preferred.

raclim commented:
I think moving forward with the first proposed solution "Move the URLs to `common/` and export it from there." would be best! Please feel free to go ahead with updating the PR and notifying us when it's done.

Hugo-RM committed:
Move p5 URLs constants to common/ directory

- Create common/p5URLs.js with centralized URL constants
- Update files to import from common/p5URLs.js
    - client/modules/IDE/hooks/useP5Version.jsx
    - client/modules/IDE/components/Preferences/index.jsx
    - server/domain-objects/createDefaultFiles.js

Hugo-RM committed:
Merge branch 'develop' into feat/url-parsing-p5.js-version

Hugo-RM committed:
Merge branch 'feat/url-parsing-p5.js-version' of https://github.com/Enochteo/p5.js-web-editor into feat/url-parsing-p5.js-version

Hugo-RM commented:
@raclim @ksen0 I updated the PR to have the URL constants moved to a new file `common/p5URLs.js`. The add-on toggle issue should now be fixed. Please let me know if there are any issues!

ksen0 commented:
Great work @Hugo-RM thanks! I've tested this and both version and add-on settings worked great.
